src/cli.rs

```diff
@@ -1,3 +1,562 @@
+            "device" => self.run_device_command(args),
+            _ => {
+                self.display_invalid_command();
+            }
+        }
+    }
+
+    fn display_message(&self, message: &str) {
+        self.console.write(message);
+    }
+
+    fn read_string(&self) -> Result<String, String> {
+        self.console.read()
+    }
+
+    fn read_number(&self) -> Result<i32, String> {
+        let input = self.console.read()?;
+        match input.trim().parse::<i32>() {
+            Ok(number) => Ok(number),
+            Err(_) => Err("Invalid number".to_string()),
+        }
+    }
+
+    fn ask_question(&self, question_type: &QuestionType, question_statement: &str) -> String {
+        match question_type {
+            QuestionType::String => self.ask_for_string(question_statement),
+            _ => panic!("Unsupported question type"),
+        }
+    }
+
+    fn ask_for_string(&self, message: &str) -> String {
+        self.display_message(message);
+        match self.read_string() {
+            Ok(answer) => answer,
+            Err(_) => self.ask_for_string(message),
+        }
+    }
+
+    fn display_help(&self) {
+        self.display_message(HELP);
+    }
+
+    fn display_version(&self) {
+        self.display_message(VERSION);
+    }
+
+    fn display_invalid_command(&self) {
+        self.display_message(INVALID_COMMAND);
+    }
+
+    fn run_device_command(&mut self, args: Vec<String>) {
+        if args.len() < 3 {
+            self.display_invalid_command();
+            return;
+        }
+
+        match args[2].as_str() {
+            "ls" | "list" => self.display_device_list(),
+            "new" => self.find_device_factory_create_new_device(args),
+            "rm" | "remove" => self.remove_device(args),
+            _ => {
+                self.display_invalid_command();
+            }
+        }
+    }
+
+    fn display_device_list(&mut self) {
+        self.display_message("Device list:");
+        let devices = self.operations.list();
+        match devices {
+            Ok(devices) => {
+                for device in devices {
+                    self.display_message(&format!("Device: {}", device.get_name()));
+                }
+            }
+            Err(e) => {
+                self.display_message(&e);
+            }
+        }
+    }
+
+    fn find_device_factory_create_new_device(&mut self, args: Vec<String>) {
+        if args.len() < 4 {
+            self.display_invalid_command();
+            return;
+        }
+        let device_key = args[3].as_str();
+        match self
+            .operations
+            .get_available_device_factories()
+            .iter()
+            .find(|&key| key.key == device_key)
+        {
+            Some(key) => self.create_new_device(key),
+            None => self.display_message("No such device configuration exists"),
+        }
+    }
+
+    fn create_new_device(&mut self, key: &DeviceFactoryKey) {
```

Comment:
Maybe you can make this function return a Result, so you have any error, you can use the "?" operator to raise it up, and at the location where you run the command, you can handle almost all errors the same way, for instance printing them in red. 

Typically you would have: 

```rs
fn create_new_device(...) -> Result<(), String> {
   ...
   let factory = self....get_device_factory(...).ok_or("No factory found with key ...")
   while(...) { ... }
   let device = factory.build()?;
   self.operations.add_device(device)?
   self.display_message("OK")
}
```

And way up, when you run command, you do the matching: 

```rs
let command_result = run_device_command(args).or_else(|e| self.display_error(e))
```